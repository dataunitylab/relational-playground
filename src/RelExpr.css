.RelExpr .clicked,
.RelExpr.clicked {
  background: #ff3;
  cursor: pointer;
}

.RelExpr .hovering,
.RelExpr.hovering {
  background: #ffffb8;
  cursor: pointer;
}
