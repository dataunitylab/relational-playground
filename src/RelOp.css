.clicked {
  background: #ff3;
  cursor: pointer;
}

.hovering {
  background: #ff3;
  cursor: pointer;
}
