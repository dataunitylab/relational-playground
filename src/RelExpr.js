// @flow
import React, {useRef} from 'react';
import {
  UnaryRelOp,
  Projection,
  Rename,
  Selection,
  BinaryRelOp,
  Except,
  Join,
  Intersect,
  Union,
} from './RelOp';
import Relation from './Relation';
import {exprToString} from './util';
import {changeExpr} from './modules/data';
import ReactDOM from 'react-dom';

import './RelExpr.css';

import type {Node, StatelessFunctionalComponent} from 'react';

type Props = {
  changeExpr: typeof changeExpr,
  expr: {[string]: any},
  ReactGA: any,
};

/** A graphical representation of a relational algebra expression */
const RelExpr: StatelessFunctionalComponent<Props> = (props) => {
  const nodeRef = useRef();

  /**
   * @param expr - a relational algebra expression object to render
   * @return a component representing the top-most expression
   */
  const buildExpr = (expr: {[string]: any}): Node => {
    // Don't try to render empty expressions
    if (!expr || Object.keys(expr).length === 0) {
      return '';
    }
    const type = Object.keys(expr)[0];
    switch (type) {
      case 'projection':
        return (
          <span>
            <UnaryRelOp
              operator={
                <Projection project={expr.projection.arguments.project} />
              }
            >
              <RelExpr
                expr={expr.projection.children[0]}
                changeExpr={props.changeExpr}
                ReactGA={props.ReactGA}
              />
            </UnaryRelOp>
          </span>
        );

      case 'selection':
        return (
          <span>
            <UnaryRelOp
              operator={
                <Selection
                  select={exprToString(expr.selection.arguments.select)}
                />
              }
            >
              <RelExpr
                expr={expr.selection.children[0]}
                changeExpr={props.changeExpr}
                ReactGA={props.ReactGA}
              />
            </UnaryRelOp>
          </span>
        );

      case 'rename':
        return (
          <span>
            <UnaryRelOp
              operator={<Rename rename={expr.rename.arguments.rename} />}
            >
              <RelExpr
                expr={expr.rename.children[0]}
                changeExpr={props.changeExpr}
                ReactGA={props.ReactGA}
              />
            </UnaryRelOp>
          </span>
        );

      case 'relation':
        return <Relation name={expr.relation} />;

      case 'except':
      case 'intersect':
      case 'join':
      case 'union':
        const operator = {
          except: <Except />,
          intersect: <Intersect />,
          join: <Join />,
          union: <Union />,
        }[type];
        return (
          <BinaryRelOp
            operator={operator}
            left={
              <RelExpr
                expr={expr[type].left}
                ReactGA={props.ReactGA}
                changeExpr={props.changeExpr}
              />
            }
            right={
              <RelExpr
                expr={expr[type].right}
                ReactGA={props.ReactGA}
                changeExpr={props.changeExpr}
              />
            }
          />
        );

      default:
        throw new Error('Invalid expression ' + JSON.stringify(expr) + '.');
    }
  };

  /**
   * @param e - the event object which generated the click
   */
  const handleExprClick = (e: SyntheticMouseEvent<HTMLElement>): void => {
    e.stopPropagation();
    const node =
      ReactDOM.findDOMNode(nodeRef.current) instanceof HTMLElement
        ? ReactDOM.findDOMNode(nodeRef.current)
        : undefined;

    if (node instanceof HTMLElement && props.changeExpr) {
      props.changeExpr(props.expr, node);
    }

    props.ReactGA.event({
      category: 'User Selecting Relational Algebra Enclosure',
      action: Object.keys(props.expr)[0],
    });
  };

  if (!props.expr || Object.keys(props.expr).length === 0) {
    return '';
  }
  const type = Object.keys(props.expr)[0];
  if (type === 'relation') {
    return (
      <span className="RelExpr" style={{margin: '.4em'}}>
        {buildExpr(props.expr)}
      </span>
    );
  } else {
    return (
      <span
        className="RelExpr"
        // $FlowFixMe[method-unbinding]
        onClick={handleExprClick}
        style={{margin: '.4em'}}
        ref={nodeRef}
      >
        {buildExpr(props.expr)}
      </span>
    );
  }
};

export default RelExpr;
