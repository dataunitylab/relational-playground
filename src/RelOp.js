// @flow
import React, {Component} from 'react';

import type {Element, Node} from 'react';

import './RelOp.css';

type Props = {
  operator: Element<any>,
  children: Node,
};

type State = {
  isHovered: boolean,
};

/** Base class for all relational algebra operators */
class RelOp extends Component<Props, State> {
  constructor() {
    super();
    this.state = {isHovered: false};
    (this: any).handleHover = this.handleHover.bind(this);
  }

  handleHover(e: SyntheticMouseEvent<HTMLElement>) {
    const hovering = e.type === 'mouseover';
    e.stopPropagation();
    this.setState(state => {
      return {...state, isHovered: hovering};
    });
  }

  render() {
    const hoverClass = 'RelOp ' + (this.state.isHovered ? 'hovering' : '');
    return (
      <span
        className={hoverClass}
        onMouseOver={this.handleHover}
        onMouseOut={this.handleHover}
      >
        {this.props.operator}({this.props.children})
      </span>
    );
  }
}

/** Projection relational algebra operator */
class Projection extends Component<{project: Array<string>}> {
  render() {
    return (
      <span>
        &pi;<sub>{this.props.project.join(',')}</sub>
      </span>
    );
  }
}

/** Rename relational algebra operator */
class Rename extends Component<{rename: {[string]: string}}> {
  render() {
    return (
      <span>
        &rho;
        <sub>
          {/* Loop over all columns to rename and combine them */}
          {Object.entries(this.props.rename)
            .map(([o, n]) => {
              return o + '/' + ((n: any): string);
            })
            .join(',')}
        </sub>
      </span>
    );
  }
}

/** Selection relational algebra operator */
class Selection extends Component<{select: Array<string>}> {
  render() {
    return (
      <span>
        &sigma;<sub>{this.props.select.join(' âˆ§ ')}</sub>
      </span>
    );
  }
}

export {RelOp as default, Projection, Rename, Selection};
