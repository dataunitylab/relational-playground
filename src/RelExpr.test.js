import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

import RelExpr from './RelExpr';
import {Selection} from './RelOp';
import {mount} from 'enzyme';

const mockStore = configureStore([]);

describe('RelExpr', () => {
  let store;

  beforeEach(() => {
    store = mockStore({data: {expr: {}}});
  });

  /** @test {RelExpr} */
  it('correctly renders a complex expression', () => {
    const expr = {
      rename: {
        arguments: {rename: {columns: {firstName: 'name'}}},
        children: [
          {
            projection: {
              arguments: {project: ['firstName', 'lastName']},
              children: [
                {
                  selection: {
                    arguments: {
                      select: {cmp: {lhs: 'salary', op: '$gt', rhs: 130000}},
                    },
                    children: [{relation: 'Doctor'}],
                  },
                },
              ],
            },
          },
        ],
      },
    };
    const tree = renderer
      .create(
        <Provider store={store}>
          <RelExpr expr={expr} changeExpr={jest.fn()} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  const condTests = [
    ['$gte', '>='],
    ['$gt', '>'],
    ['$lt', '<'],
    ['$lte', '<='],
    ['$ne', '!='],
    ['$eq', '='],
  ];

  /** @test {RelExpr} */
  it.each(condTests)('it correctly renders a %s condition as %s', (op, str) => {
    const expr = {
      selection: {
        arguments: {select: {cmp: {lhs: 'salary', op: op, rhs: 130000}}},
        children: [{relation: 'Doctor'}],
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <RelExpr expr={expr} changeExpr={jest.fn()} />
      </Provider>
    );
    expect(wrapper.find(Selection).text()).toContain(
      'salary ' + str + ' 130000'
    );
  });

  /** @test {RelExpr} */
  it('produces an error for an invalid expression', () => {
    const errorObject = console.error;
    console.error = jest.fn();

    expect(() => {
      renderer.create(
        <Provider store={store}>
          <RelExpr expr={{invalidExpr: 42}} changeExpr={jest.fn()} />
        </Provider>
      );
    }).toThrow();

    console.error = errorObject;
  });

  /** @test {RelExpr} */
  it('doesnt change the expression when clicked and relation', () => {
    const mockAction = jest.fn();
    const mockEvent = jest.fn();
    const expr = {relation: 'foo'};
    const wrapper = mount(
      <Provider store={store}>
        <RelExpr
          ReactGA={{event: mockEvent}}
          expr={expr}
          changeExpr={mockAction}
        />
      </Provider>
    );

    // Click on the expression
    wrapper.simulate('click', {stopPropagation: jest.fn()});

    // An action changing the expression should fire
    // Should not have action attached when type relation
    expect(mockAction.mock.calls.length).toBe(0);

    // Don't expect analytics on a relation
    expect(mockEvent.mock.calls.length).toBe(0);
  });

  /** @test {RelExpr} */
  it('changes the expression when clicked not relation', () => {
    const mockAction = jest.fn();
    const mockEvent = jest.fn();
    const expr = {
      selection: {
        arguments: {
          select: 'foo',
        },
        children: [{relation: 'bar'}],
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <RelExpr
          ReactGA={{event: mockEvent}}
          expr={expr}
          changeExpr={mockAction}
        />
      </Provider>
    );

    // Click on the expression
    wrapper.simulate('click', {stopPropagation: jest.fn()});

    // An action changing the expression should fire
    expect(mockAction.mock.calls.length).toBe(1);
    expect(mockAction.mock.calls[0][0]).toBe(expr);

    // And also an analytics event
    expect(mockEvent.mock.calls.length).toBe(1);
    expect(mockEvent.mock.calls[0][0].category).toBe(
      'User Selecting Relational Algebra Enclosure'
    );
    expect(mockEvent.mock.calls[0][0].action).toBe('selection');
  });

  /** @test {RelExpr} */
  it('should add (and remove) a class on hover', () => {
    const mockAction = jest.fn();
    const mockEvent = jest.fn();
    const expr = {
      selection: {
        arguments: {
          select: 'foo',
        },
        children: [{relation: 'bar'}],
      },
    };
    const wrapper = mount(
      <Provider store={store}>
        <RelExpr
          ReactGA={{event: mockEvent}}
          expr={expr}
          changeExpr={mockAction}
        />
      </Provider>
    );

    // Hovering class should be off by default
    expect(wrapper.hasClass('hovering')).toBeFalsy();

    // Hovering should add the class and not propagate the event
    const mockStop = jest.fn();
    wrapper.simulate('mouseover', {
      type: 'mouseover',
      stopPropagation: mockStop,
    });
    expect(wrapper.exists('.hovering')).toBeTruthy();
    expect(mockStop.mock.calls.length).toBe(1);

    // Mouse out should remove the class
    wrapper.simulate('mouseout', {
      type: 'mouseout',
      stopPropagation: jest.fn(),
    });
    expect(wrapper.exists('.hovering')).toBeFalsy();
  });
});
