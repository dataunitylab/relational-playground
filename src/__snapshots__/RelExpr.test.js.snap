// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelExpr correctly renders a complex expression 1`] = `
<span
  className="RelExpr"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "margin": ".4em",
    }
  }
>
  <span>
    <span>
      <span>
        <span>
          ρ
          <sub>
            firstName/name
          </sub>
        </span>
        (
        <span
          className="RelExpr"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "margin": ".4em",
            }
          }
        >
          <span>
            <span>
              <span>
                <span>
                  π
                  <sub>
                    firstName,lastName
                  </sub>
                </span>
                (
                <span
                  className="RelExpr"
                  onClick={[Function]}
                  onMouseOut={[Function]}
                  onMouseOver={[Function]}
                  style={
                    Object {
                      "margin": ".4em",
                    }
                  }
                >
                  <span>
                    <span>
                      <span>
                        <span>
                          σ
                          <sub>
                            salary &gt; 130000
                          </sub>
                        </span>
                        (
                        <span
                          className="RelExpr"
                          style={
                            Object {
                              "margin": ".4em",
                            }
                          }
                        >
                          <span
                            className="Relation"
                            data-testid="span"
                          >
                            Doctor
                          </span>
                        </span>
                        )
                      </span>
                    </span>
                  </span>
                </span>
                )
              </span>
            </span>
          </span>
        </span>
        )
      </span>
    </span>
  </span>
</span>
`;
